---
title: "Getting Started with RefineBio"
author:
- name: First Author
  affiliation: First Author's Affiliation
- name: Second Author
  affiliation: Second Author's Affiliation
  email: corresponding@author.com
package: packageName
output:
  BiocStyle::html_document
abstract: |
  Description of your vignette
vignette: |
  %\VignetteIndexEntry{Getting started with RefineBio}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The `RefineBio` package provides an R/Bioconductor interface to the 
[RefineBio API](https://api.refine.bio/). RefineBio provides
a web service for downloading gene expression data from a [NCBI GEO](https://www.ncbi.nih.gov/geo)
and [NCBI SRA](https://www.ncbi.nih.gov/sra). The package 
provides functions for downloading data from RefineBio and 
loading it into R, akin to the `GEOquery`[https://bioconductor.org/packages/GEOquery] package.

Advantages of using RefineBio over GEO data or SRA directly are:

1. RefineBio provides homogeneously processed data for several
   platforms of microarray. While homogeous processing doe::: notes
   not guarantee that the data is comparable, it does remove
   some of the variability that can be introduced by different
   processing pipelines.
2. RefineBio provides homogeneously processed data for RNA-seq,
   including quantification of gene expression and (optional) 
   normalization.
3. RefineBio has done some light curation of sample metadata,
   thus providing a few columns of metadata that are standardized
   across datasets.

Advantages of using the `RefineBio` package over the web interface
are:

1. The `RefineBio` package provides a reproducible way to download
    data. The web interface is subject to change, and the `RefineBio`
    package provides a stable interface to the API.
2. The `RefineBio` package provides the opportunity to download
    data in a programmatic way, which can be useful for batch
    downloading and processing data.

   :::



# Installation

Installation is straightforward using the `BiocManager` package:

    BiocManager::install('seandavi/RefineBio')

# Basic usage

Load the library:

```{r}
library(RefineBio)
library(SummarizedExperiment)
```

Similar to the `GEOquery` package, the `RefineBio` package provides
a single function that can be used to import data into R as 
a `SummarizedExperiment` object. 

Peruse the [RefineBio website](https://www.refine.bio/) to find an 
accession of interest and then plug it into the `get_refinebio()`
function. For example, to download the data from the accession
`GSE37382`:

```{r message=FALSE}
se <- get_refinebio("SRP127409")
```

Like `GEOquery`, the `RefineBio` package provides a **LIST** of 
`SummarizedExperiment` objects.

The `SummarizedExperiment` objects can then be used as you
would any other `SummarizedExperiment` object. For example, to
plot the first 10 genes:

```{r}
se[[1]]
```

Sample metadata arrive and are parsed into the usual `colData`.

```{r}
head(colData(se[[1]]))
```

And to see the columns more clearly:

```{r}
colnames(colData(se[[1]]))
```

Additional study-level metadata (that also includes 
the sample metadata) is also loaded and stored in the
`metadata` slot.

```{r}
head(metadata(se[[1]])$experimental_details)
```

# Citations

Cite the refine.bio project as:

> Casey S. Greene, Dongbo Hu, Richard W. W. Jones, Stephanie Liu, David S. Mejia, Rob Patro, Stephen R. Piccolo, Ariel Rodriguez Romero, Hirak Sarkar, Candace L. Savonen, Jaclyn N. Taroni, William E. Vauclain, Deepashree Venkatesh Prasad, Kurt G. Wheeler. refine.bio: a resource of uniformly processed publicly available gene expression datasets. URL: https://www.refine.bio

And the `RefineBio` package as:

> Alexander Ho and Sean Davis. RefineBio: An R/Bioconductor interface to the refine.bio API. URL: https://github.com/seandavi/RefineBio



# Status

This project is undergoing active development, but basic functionality for 
loading an experiment from RefineBio exists.

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
